subprojects/core-api/src/main/java/org/gradle/api/tasks/TaskInputFilePropertyBuilder.java

```diff
@@ -69,6 +69,13 @@
     TaskInputFilePropertyBuilder withPathSensitivity(PathSensitivity sensitivity);
 
     /**
+     * Sets the normalizer to use for this property.
+     *
+     * @since 4.3
+     */
+    TaskInputFilePropertyBuilder withNormalizer(Class<? extends PropertyNormalizer> snapshotter);
+
```

Comment:
Did you consider using the annotations here? That would mean less public types.

Reply:
The reason for not using the annotations is mainly discoverability. If we had a method like `treadAsIfAnnotatedWith()`, what would you pass it? What should happen if I called it with  `@Classpath`? Okay, we know this one. But what if I passed `@Optional` or `@PathSensitive`? How about `@Override` or `@Nonnull`?

Using an explicit interface is much more precise, limiting the scope to what we want to handle, and that only. We also build on the concept of input normalization that we've introduced with runtime classpath normalization a while back.

Reply:
Well, it creates inconsistency. We could do it the other way around too though and introduce `@NormalizeWith`. I just think it is confusing to have two types (an annotation and a marker interface) representing the same thing.

Reply:
The semantics of the two APIs won't be the same, and that's okay. What we want is for them to be equally expressive/powerful.

Putting `@Classpath` on a task property means several things that would be expressed in different ways using the runtime API:

- it's an input to the task; with the runtime API this fact is captured by calling `task.inputs.files()` as opposed to `task.outputs.files()`
- it's a bunch of files; which is expressed by calling `task.inputs.files()`, not `task.inputs.property()`
- it has some special normalization (we treat top-level elements as `PathSensitivity.NONE` while retaining order, top-level directories and ZIP/JAR entries are exploded, and elements inside the resulting trees are handled as `PathSensitivity.RELATIVE`, ignoring order); this is represented by `task.inputs.files(...).withNormalizer(ClasspathNormalizer)`.

Reply:
Actually, we don't even want the two APIs to be equally powerful/expressive. What we want is for the runtime API to be at least as expressive/powerful as the annotations API.

Reply:
I get what you mean now. Using the classpath normalizer is just one effect of adding the `@Classpath` annotation.